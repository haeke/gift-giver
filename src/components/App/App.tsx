import React from "react";
import { Button } from "react-bootstrap";

interface Gift {
  id: number;
}

interface State {
  gifts: Array<Gift | any>;
}

class App extends React.Component<{}, State> {
  state = {
    gifts: []
  };

  addGift = () => {
    const { gifts } = this.state;
    // get a list of all of the ids inside the gifts array
    let ids = gifts.map(gift => gift.id);
    // get the maximum id inside of the ids array account for the possiblity of there being no items
    let maxId = ids.length > 0 ? Math.max(...ids) : 0;

    this.setState({
      gifts: [...gifts, { id: maxId + 1 }]
    });
  };
  render() {
    return (
      <div>
        <h1>App Test SnapShot</h1>
        <Button className="btn-add" onClick={this.addGift}>
          Add
        </Button>
      </div>
    );
  }
}

export default App;
