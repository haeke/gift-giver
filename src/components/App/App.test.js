import React from "react";
// Shallow renders the root node and returns a shallow wrapper around it. It must be a single-node wrapper.
//
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import App from "./App";

// required to make enzyme work
configure({ adapter: new Adapter() });

const app = shallow(<App />);

it("expects app to work", () => {
  expect(app).toMatchSnapshot();
});

it("initializes the `state` to contain an empty array of gifts", () => {
  // checks the state object inside of the app component for an object called gift. The gift object should contain an empty array.
  expect(app.state().gifts).toEqual([]);
});

it("adds a new gift to the `state` by clicking the add gift button", () => {
  // use the find method to look for the react component that contains the class .btn-add
  // use the simulate function to mock user interaction
  app.find(".btn-add").simulate("click");
  //
  expect(app.state().gifts).toEqual([{ id: 1 }]);
});

it("adds a new gift to the rendered list when clicking the add gift button", () => {
  // use the find method
  // use the simulate function to mock user interaction
  app.find(".btn-add").simulate("click");
  // find the dom node with gift-list then look at all of its children nodes and check that the length is equal 1.
  expect(app.find(".gift-list").children().length).toEqual(2);
});
