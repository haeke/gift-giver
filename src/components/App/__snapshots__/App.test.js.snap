// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expects app to work 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          App Test SnapShot
        </h1>,
        <ForwardRef
          active={false}
          className="btn-add"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          Add
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "App Test SnapShot",
        },
        "ref": null,
        "rendered": "App Test SnapShot",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "children": "Add",
          "className": "btn-add",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
          "variant": "primary",
        },
        "ref": null,
        "rendered": "Add",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "active": false,
            "disabled": false,
            "type": "button",
            "variant": "primary",
          },
          "displayName": "Button",
          "render": [Function],
        },
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            App Test SnapShot
          </h1>,
          <ForwardRef
            active={false}
            className="btn-add"
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="primary"
          >
            Add
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "App Test SnapShot",
          },
          "ref": null,
          "rendered": "App Test SnapShot",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
            "children": "Add",
            "className": "btn-add",
            "disabled": false,
            "onClick": [Function],
            "type": "button",
            "variant": "primary",
          },
          "ref": null,
          "rendered": "Add",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "active": false,
              "disabled": false,
              "type": "button",
              "variant": "primary",
            },
            "displayName": "Button",
            "render": [Function],
          },
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
